src/pkg/cached/project/redis/manager.go

```diff
@@ -0,0 +1,198 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	libcache "github.com/goharbor/harbor/src/lib/cache"
+	"github.com/goharbor/harbor/src/lib/config"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/lib/q"
+	"github.com/goharbor/harbor/src/lib/retry"
+	"github.com/goharbor/harbor/src/pkg/cached"
+	"github.com/goharbor/harbor/src/pkg/project"
+	"github.com/goharbor/harbor/src/pkg/project/models"
+)
+
+var _ CachedManager = &manager{}
+
+// CachedManager is the interface combines raw resource manager and cached manager for better extension.
+type CachedManager interface {
+	// Manager is the raw resource manager.
+	project.Manager
+	// Manager is the common interface for resource cache.
+	cached.Manager
+}
+
+// manager is the cached manager implemented by redis.
+type manager struct {
+	// delegator delegates the raw crud to DAO.
+	delegator project.Manager
+	// client returns the redis cache client.
+	client func() libcache.Cache
+	// keyBuilder builds cache object key.
+	keyBuilder *cached.ObjectKey
+	// lifetime is the cache life time.
+	lifetime time.Duration
+}
+
+// NewManager returns the redis cache manager.
+func NewManager(m project.Manager) *manager {
+	return &manager{
+		delegator:  m,
+		client:     func() libcache.Cache { return libcache.Default() },
+		keyBuilder: cached.NewObjectKey(cached.ResourceTypeProject),
+		lifetime:   time.Duration(config.CacheExpireHours()) * time.Hour,
+	}
+}
+
+func (m *manager) Create(ctx context.Context, project *models.Project) (int64, error) {
+	return m.delegator.Create(ctx, project)
+}
+
+func (m *manager) Count(ctx context.Context, query *q.Query) (total int64, err error) {
+	return m.delegator.Count(ctx, query)
+}
+
+func (m *manager) List(ctx context.Context, query *q.Query) ([]*models.Project, error) {
+	return m.delegator.List(ctx, query)
+}
+
+func (m *manager) ListRoles(ctx context.Context, projectID int64, userID int, groupIDs ...int) ([]int, error) {
+	return m.delegator.ListRoles(ctx, projectID, userID, groupIDs...)
+}
+
+func (m *manager) Delete(ctx context.Context, id int64) error {
+	p, err := m.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	// pass on delete operation
+	if err := m.delegator.Delete(ctx, id); err != nil {
+		return err
+	}
+	// clean cache
+	m.cleanUp(ctx, p)
+	return nil
+}
+
+func (m *manager) Get(ctx context.Context, idOrName interface{}) (*models.Project, error) {
+	var (
+		key string
+		err error
+	)
+
+	id, ok := idOrName.(int64)
```

Comment:
update & leverage the common/util.ParseProjectIDOrName()?

Reply:
updated